docs/eventsV2.md

```diff
@@ -0,0 +1,268 @@
+is an error.
+
+If there is an execution error, the QueueWorker will immediately tell PubSub to retry
+in 5 minutes.
+
+If there was an operational error, the QueueWorker will not be able to do anything to
+the event. However, the QueueWorker will not acknowledge the PubSub notification, and
+so when the PubSub ack deadline for that notification passes, PubSub will attempt to
+deliver it again. PubSub is configured to try a message 5 times, then it publishes it
+in a deadletter queue. We do not currently have any automation on the deadletter
+queue.
+
+An event that has more than 5 delays will be deleted by the QueueWorker.
+
+"Bad" notifications (perhaps they can't be read by the queueworker) will be discarded
+by PubSub after a week.
+
+### Queue counts
+
+**FSTODO**: add an index
+We do a DB query for the number of events for that canvas/handler.
+
+### Emit
+
+Done in `LibEvent` via `emit`, or automatically via `CronChecker`. Calls
+`EventQueueV2.enqueue`. This adds a new value to the events table with:
+
+- `locked_at = NULL`
+- `enqueued_at = CURRENT_TIMESTAMP`
+
+Note that `CronChecker` does not use `events_v2` table information for Cron
+scheduling (eg determining if it has been an hour since the last Cron event). It does
+that separately in the `cron_records` table and just emits events on the appropriate
+schedule, leaving it to the queues to handler after that.
+
+Emit also adds a notification to the PubSub topic. This will be delivered to a
+QueueWorker to tell it to try fetching and running an event. After 5 retries, PubSub
+will drop the notification, sending it to a dead-letter queue (which at the moment
+has no UI or anything).
+
+### QueueWorker Execution
+
+`EventQueueV2.dequeue` fetches a notification from PubSub and runs the process to
+execute it. First it will check if it should run it, looking at retries, whether
+another worker has set `locked_at` and holds the lock, whether scheduling rules tell
+us not to run it, whether it's not time to run it yet, or if the event is missing.
+
+Notifications have a built-in _ack deadline_ - they must be _acknowledged_ within a
+changeable time limit. The default is 60 seconds. If a queueworker decides not to run
+an event immediately, it will put back the notification by setting the ack deadline
+to a more appropriate time. If it decides it should not be run (eg if paused) it will
+remove it from PubSub by _acknowledging_ it.
+
+If a QueueWorker decides to runs the event, it will increase the _ack time_ to 5
+minutes to give it time to execute. When it is done, it removes it from the queue by
+_acknowledging_ it.
+
+Because the event is just a notification, it's basically almost always fine to put it
+back in so long as the DB row for the event exists (which means it's done). However,
+once it has been delivered 5 times, PubSub will stop trying to retry it.
+
+Note that normally, only one notification exists for an event. The exception is when
+workers are unpaused.
+
+```mermaid
```

Comment:
I don't have great detailed feedback at this point, but don't find this chart/graph particularly useful or digest-able. Could just be me.

Reply:
The intent is for this chart to describe the actual flow, while the code tries to implement the chart.

I've found it very helpful while developing the algorithm, as it's much more high level than code and easier to see flaws in the logic.